File: CS/DxSample.Service/Controllers/OrdersController.cs
using DxSample.Service.Models;
using System.Linq;
using System.Web.Http;
using System.Threading.Tasks;
using System.Net;
using System.Web.OData;

namespace DxSample.Service.Controllers {
    public class OrdersController : ODataController {
        private bool OrderExists(int key) {
            return DataContext.Orders.Any(c => c.ID == key);
        }

        [EnableQuery]
        public IQueryable<Order> Get() {
            return DataContext.Orders;
        }

        [EnableQuery]
        public SingleResult<Order> Get([FromODataUri] int key) {
            IQueryable<Order> result = DataContext.Orders.Where(c => c.ID == key);
            return SingleResult.Create(result);
        }

        public async Task<IHttpActionResult> Post(Order order) {
            if (!this.ModelState.IsValid) return this.BadRequest(this.ModelState);
            order = await DataContext.AddOrder(order.Name, order.CustomerID);
            return this.Created(order);
        }

        public async Task<IHttpActionResult> Patch([FromODataUri] int key, Delta<Order> order) {
            if (!this.ModelState.IsValid) return this.BadRequest(this.ModelState);
            Order entity = await DataContext.FindOrderAsync(key);
            if (entity == null) return this.NotFound();
            order.Patch(entity);
            return this.Updated(entity);
        }

        public async Task<IHttpActionResult> Put([FromODataUri] int key, Order update) {
            if (!this.ModelState.IsValid) return this.BadRequest(this.ModelState);
            if (key != update.ID) return this.BadRequest();
            await DataContext.UpdateOrder(update);
            return this.Updated(update);
        }

        public async Task<IHttpActionResult> Delete([FromODataUri] int key) {
            Order order = await DataContext.FindOrderAsync(key);
            if (order == null) return this.NotFound();
            await DataContext.RemoveOrder(order);
            return this.StatusCode(HttpStatusCode.NoContent);
        }
    }
}

File: VB/DxSample.Service/Controllers/OrdersController.vb
Imports DxSample.Service.Models
Imports System.Linq
Imports System.Web.Http
Imports System.Threading.Tasks
Imports System.Net
Imports System.Web.OData

Namespace DxSample.Service.Controllers
	Public Class OrdersController
		Inherits ODataController

		Private Function OrderExists(ByVal key As Integer) As Boolean
			Return DataContext.Orders.Any(Function(c) c.ID = key)
		End Function

		<EnableQuery>
		Public Function [Get]() As IQueryable(Of Order)
			Return DataContext.Orders
		End Function

		<EnableQuery>
		Public Function [Get](<FromODataUri> ByVal key As Integer) As SingleResult(Of Order)
			Dim result As IQueryable(Of Order) = DataContext.Orders.Where(Function(c) c.ID = key)
			Return SingleResult.Create(result)
		End Function

		Public Async Function Post(ByVal order As Order) As Task(Of IHttpActionResult)
			If Not Me.ModelState.IsValid Then
				Return Me.BadRequest(Me.ModelState)
			End If
			order = Await DataContext.AddOrder(order.Name, order.CustomerID)
			Return Me.Created(order)
		End Function

		Public Async Function Patch(<FromODataUri> ByVal key As Integer, ByVal order As Delta(Of Order)) As Task(Of IHttpActionResult)
			If Not Me.ModelState.IsValid Then
				Return Me.BadRequest(Me.ModelState)
			End If
			Dim entity As Order = Await DataContext.FindOrderAsync(key)
			If entity Is Nothing Then
				Return Me.NotFound()
			End If
			order.Patch(entity)
			Return Me.Updated(entity)
		End Function

		Public Async Function Put(<FromODataUri> ByVal key As Integer, ByVal update As Order) As Task(Of IHttpActionResult)
			If Not Me.ModelState.IsValid Then
				Return Me.BadRequest(Me.ModelState)
			End If
			If key <> update.ID Then
				Return Me.BadRequest()
			End If
			Await DataContext.UpdateOrder(update)
			Return Me.Updated(update)
		End Function

		Public Async Function Delete(<FromODataUri> ByVal key As Integer) As Task(Of IHttpActionResult)
			Dim order As Order = Await DataContext.FindOrderAsync(key)
			If order Is Nothing Then
				Return Me.NotFound()
			End If
			Await DataContext.RemoveOrder(order)
			Return Me.StatusCode(HttpStatusCode.NoContent)
		End Function
	End Class
End Namespace